### ISSUE TITLE
Add support for a transition argument to power{On,Off,Toggle}

### ISSUE THREAD

**Issue by lopter** [lopter/lightsd#5]
Equivalent to:

    prev_color = get_light_state(target)
    set_light_from_hsbk(target, 0, 0, 0 or 1, temperature, 60 * 1000 * 5)
    sleep(60 * 5)
    powerOff(target)
    set_light_from_hsbk(target, prev_color, 0) # requires firmware >= 1.5 for the Original 1000

The LIFX REST API doesn't seem to update the state (color) of the bulb during the transition, but this will.

On top of my head:

- add new optional argument to the power functions;
- wrap something around [lgtd_timer](https://github.com/lopter/lightsd/blob/349773d2d081108c2124f023d302bcc28c7efbb8/core/timer.h).

**Comment by vegansbane** [comment]
Thanks lopter, this will help free me from cloud dependence with my use case!

**Comment by chendo** [comment]
You could use `Light::SetPower` (http://lan.developer.lifx.com/docs/light-messages#section-setpower-117) and specify the duration there so you don't need to use `Light::Set` at all

**Comment by lopter** [comment]
Hello @chendo,

Sorry I never followed-up on that, it's a very good tip and I see it's supported as far as back as firmware 1.5 of the Original 1000!

Unfortunately, the state of the bulb doesn't get updated with this command (i.e: the brightness isn't updated and power is set toggled immediately), so I'll still emulate the feature as I described.

It's actually cool, it will make me add an internal effects API in lightsd that I'll leverage to implement other features.

**Comment by chendo** [comment]
Ah, as you want it to block the method until he transition completes?

**Comment by lopter** [comment]
I'm not sure what you mean by block, I just want the state of the bulb to reflect the transition in real time.

**Comment by chendo** [comment]
As in you want the physical state of the bulb or what LightState reports to reflect the transition in real time?

**Comment by lopter** [comment]
Yes!

**Comment by chendo** [comment]
Yes to which one?

**Comment by lopter** [comment]
I mean, yes as in LightState should reflect the transition in real time.

**Comment by chendo** [comment]
Power transitions are reflected through the `power` UINT16 in `LightState` if that helps?

**Comment by lopter** [comment]
Ah I [did notice that](https://github.com/lopter/lightsd/commit/da9c1bb7e38e10c8122338df50a214b289d0a040) however I'm clamping the value to 0 or 0xffff (I think I have seen some LIFX code do the same thing) since the documentation explicitly says only those values are [supported](http://lan.developer.lifx.com/v2.0/docs/device-messages#section-power-level).

How is this value related to the brightness? Is there a (mathematical) formula/function to convert between the two?

Thanks for the help!

**Comment by chendo** [comment]
Yeah, you can only set 0 for off and non-zero for on for power. I'd have to check the code to be sure but I assume the brightness is multiplied by the percentage of power. If you want to handle transitions nicely, I have logic in the HTTP API that does stuff like set a 0-brightness version of the desired colour before turning the bulb on, then do the transition to the desired colour to fade on properly, so you can do something similar if you want to deal with that nicely.

**Comment by lopter** [comment]
Hmm, I think I get it, SetPower is just misnamed, it should be named SetBrightness am I correct?

**Comment by lopter** [comment]
Ok, here is how I get it:

SetPower is indeed SetPower and works in a boolean way.

However StatePower does uses the full 0-65535 range, and its value isn't related to the brightness or the power level, but is instead the transition: if I do SetPower with a 4s transition at *t* 0s StatePower will be 0 at *t* 2s StatePower will be 32767 and at *t* 4s StatePower will be 65535 not matter what the targeted brightness is.

So I'll keep doing what I [was saying](https://github.com/lopter/lightsd/issues/5#issuecomment-167445728) because it will naturally reflect the absolute brightness value and I'm already sampling the bulbs with [Get](http://lan.developer.lifx.com/docs/light-messages#section-get-101).

**Comment by chendo** [comment]
That'll work assuming that the user exclusively only uses lightsd for control. The HTTP API and the app does use SetPower with transitions where applicable as to preserve the brightness level before off.

**Comment by lopter** [comment]
No worries, lightsd will implement it in a correct way, you can actually see my work in progress in the lightsd-mq repository.
